### PR TITLE
Speed up by avoiding `arguments` and apply

### PR DESCRIPTION
**Pull request by RubenVerborgh**
This pull request avoids passing `arguments` around, which is [bad for performance](https://github.com/petkaantonov/bluebird/wiki/Optimization-killers#32-leaking-arguments). It also avoids unnecessary `apply` calls by optimizing for common numbers of parameters, [as in Node.js](https://github.com/nodejs/node/blob/v6.4.0/lib/timers.js#L335).

### CODE REVIEW

File: setImmediate.js [hunk 1]
```diff
@@ -9,24 +9,21 @@
...
-
-    function addFromSetImmediateArguments(args) {
-        tasksByHandle[nextHandle] = partiallyApplied.apply(undefined, args);
-        return nextHandle++;
-    }
-
-    // This function accepts the same arguments as setImmediate, but
-    // returns a function that requires no arguments.
-    function partiallyApplied(handler) {
-        var args = [].slice.call(arguments, 1);
```
Comment (RubenVerborgh) [inline setImmediate.js]: [Performance killer](https://github.com/petkaantonov/bluebird/wiki/Optimization-killers#32-leaking-arguments)

**Comment by domenic** [comment]
Benchmarks before and after, please.

File: setImmediate.js [hunk 2]
```diff
@@ -35,13 +32,13 @@
...
             // "too much recursion" error.
-            setTimeout(partiallyApplied(runIfPresent, handle), 0);
+            setTimeout(function () { runIfPresent(handle); }, 0);
         } else {
             var task = tasksByHandle[handle];
             if (task) {
                 currentlyRunningATask = true;
                 try {
-                    task();
+                    run(task);
```
Comment (RubenVerborgh) [inline setImmediate.js]: Separate function, outside of `try finally` [for performance](https://github.com/petkaantonov/bluebird/wiki/Optimization-killers#2-unsupported-syntax).

**Comment by RubenVerborgh** [comment]
Benchmarks for the same case in Node.js are [here](https://github.com/nodejs/node/pull/406); will add some for this instance as well.

**Comment by RubenVerborgh** [comment]
[Benchmark](https://gist.github.com/a9240a4a0db280f597a16fc3ed5411bd)

|                       | benchmark time (s) |
|-----------------------|-------------------:|
| native `setImmediate` |                7.3 |
| current `master`      |               16.2 |
| this pull request     |                2.3 |

**Comment by domenic** [comment]
Awesome. Will merge and publish some time when I'm not at work.

File: setImmediate.js [hunk 3]
```diff
@@ -9,24 +9,21 @@
...
-                handler.apply(undefined, args);
-            } else {
-                (new Function("" + handler))();
-            }
-        };
+    var registerImmediate;
+
+    function setImmediate(callback) {
+      // Callback can either be a function or a string
+      if (typeof callback !== 'function')
```
Comment (calvinmetcalf) [inline setImmediate.js]: curly braces around the body please

File: setImmediate.js [hunk 4]
```diff
@@ -9,24 +9,21 @@
...
+        callback = new Function('' + callback);
+      // Copy function arguments
+      var args = new Array(arguments.length - 1);
+      for (var i = 0; i < args.length; i++)
```
Comment (calvinmetcalf) [inline setImmediate.js]: curly braces

File: setImmediate.js [hunk 5]
```diff
@@ -35,13 +32,13 @@
         if (currentlyRunningATask) {
             // Delay by doing a setTimeout. setImmediate was tried instead, but in Firefox 7 it generated a
             // "too much recursion" error.
-            setTimeout(partiallyApplied(runIfPresent, handle), 0);
+            setTimeout(function () { runIfPresent(handle); }, 0);
```
Comment (calvinmetcalf) [inline setImmediate.js]: couldn't we just do `setTimeout(runIfPresent, 0, handle);`

File: setImmediate.js [hunk 6]
```diff
@@ -9,24 +9,21 @@
...
+    function setImmediate(callback) {
+      // Callback can either be a function or a string
+      if (typeof callback !== 'function')
+        callback = new Function('' + callback);
+      // Copy function arguments
+      var args = new Array(arguments.length - 1);
+      for (var i = 0; i < args.length; i++)
+          args[i] = arguments[i + 1];
+      // Store and register the task
+      var task = { callback: callback, args: args };
```
Comment (calvinmetcalf) [inline setImmediate.js]: we should probably make an object here with a constructor [like it do in node-process](https://github.com/defunctzombie/node-process/blob/1c4943a088579d57cc82876e6ed3972fd2595e84/browser.js#L127-L134)

File: setImmediate.js (hunk 2 above)
Comment (calvinmetcalf) [inline setImmediate.js]: ha fun fact from testing, try/finally is not an optimization killer...in v8, it is one in firefox

File: setImmediate.js (hunk 6 above)
Comment (RubenVerborgh) [inline setImmediate.js]: Tested, but there's no performance advantage. v8 has enough with the fixed object assignment to optimize it.

**Comment by RubenVerborgh** [comment]
Updated code, passes `npm lint` now.

**Pull request merged**